src/slacko.ml

```diff
@@ -765,33 +765,22 @@ let groups_list ?exclude_archived session =
       (match d |> groups_list_obj_of_yojson with
         | Result.Ok x -> `Success x.groups
         | Result.Error x -> `ParseFailure x)
-    | #bot_error
     | #parsed_auth_error as res -> res
     | _ -> `Unknown_error
 
-exception No_matches
-exception No_unique_matches
-exception Lookup_failed
-
 (* look up the id of query from results provided by the listfn *)
 let lookupk session listfn filterfn k =
   match%lwt listfn session with
-  | `Success channels -> (match List.filter filterfn channels with
-    | [] -> Lwt.fail No_matches
-    | [x] -> Lwt.return @@ k x
-    | _ -> Lwt.fail No_unique_matches)
-  | _ -> Lwt.fail Lookup_failed
+  | #parsed_auth_error as e -> Lwt.return e
+  | `Success items -> Lwt.return @@ k @@ List.filter filterfn items
```

Comment:
The first thing is all about types, and this is one of those cases where it's probably better to make them explicit instead of inferring them. I'll do that, but first I'll explain what's going on.

TL;DR: The pattern match must be exhaustive or the compile fails with a type error.

Here's the inferred type of `lookupk` (according to merlin):
```
'a ->
('a ->
 [< `Account_inactive
  | `Invalid_auth
  | `Not_authed
  | `ParseFailure of string
  | `Success of 'b list
  | `Unhandled_error of string
  | `Unknown_error ]
 Lwt.t) ->
('b -> bool) -> ('b list -> ([> parsed_auth_error ] as 'c)) -> 'c Lwt.t
```
The first arg is a `session`, but we don't care as long as we can pass it to `listfn`. The second arg is `listfn`. We give it the session (`'a`) and we `match%lwt` its return value, so its return value is inferred to be (a subset of) the variants we know how to handle (wrapped in `Lwt.t`). That gives us the set of allowed errors (from the `| #parsed_auth_error as e ->` match) and `` `Success of 'b list ``. We don't care what `'b` is, as long as we can pass it to `filterfn` (the third arg with type `'b -> bool`) and `k` (the fourth arg).

`k` is inferred to have the type `'b list -> ([> parsed_auth_error ] as 'c)`, where `'c` is also `lookupk`'s return type` because we return whatever value it gives us. Because we can also return a `parsed_auth_error` from the other match, `k`'s return type must be compatible with that. However, `'c` includes the "not found" error value (either `` `Channel_not_found `` or `` `User_not_found ``) and that must remain opaque to `lookupk` so we can use it for both channels and users without adding superfluous errors to their return types.

I can (and plan to) define `type 'a listfn = session -> [`Success of 'a list | parsed_auth_error] Lwt.t` and annotate `lookupk`'s `listfn` arg to make it explicit (and make the inferred types more readable), but that's essentially what's being inferred already anyway.